src/manifest/pfm.rs

```diff
@@ -0,0 +1,497 @@
+
+use core::mem;
+
+use zerocopy::AsBytes;
+use zerocopy::FromBytes;
+
+use crate::crypto::sha256;
+use crate::crypto::sha256::Hasher as _;
+use crate::hardware::flash::Flash;
+use crate::hardware::flash::FlashExt as _;
+use crate::hardware::flash::FlashIo;
+use crate::io::Read as _;
+use crate::manifest::container::Container;
+use crate::manifest::provenance;
+use crate::manifest::Error;
+use crate::mem::Arena;
+use crate::protocol::wire::WireEnum as _;
+
+wire_enum! {
+    /// A PFM element type.
+    pub enum ElementType: u8 {
+      /// A `FlashDevice` defines global information pertinent to the entire
+      /// flash device a particular PFM describes.
+      FlashDevice = 0x00,
+
+      /// A list of firmware versions (i.e., contents of flash) allowed by
+      /// policy on the flash device a particular PFM describes.
+      AllowableFw = 0x01,
+
+      /// A firmware version, as a subelement of an `AllowableFw`.
+      FwVersion = 0x02,
+
+      /// A bytestring identifier for the platform that this PFM describes.
+      PlatformId = 0x03,
+    }
+}
+
+wire_enum! {
+    /// A hash type, as defined in certain parts of the PFM.
+    ///
+    /// Note that we currently only support the SHA-256 variant, even though
+    /// Cerberus permits SHA-384 and SHA-512 as well.
+    #[allow(missing_docs)]
+    pub enum HashType: u8 {
+      Sha256 = 0b000,
+      // Sha384 = 0b001,
+      // Sha512 = 0b010,
+    }
+}
+
+/// An entry to a PFM's table of contents.
+///
+/// A TOC entry describes an element in the PFM, such as its format and its
+/// location.
+///
+/// A TOC entry is encoded exactly the same way as this struct is laid
+/// out, byte-for-byte.
+///
+/// See [`Toc`](struct.Toc.html).
+#[derive(Copy, Clone, PartialEq, Eq, AsBytes, FromBytes)]
+#[repr(C)]
+pub struct TocEntry {
+    element_type: u8,
+    format_version: u8,
+    offset: u16,
+    len: u16,
+    parent_idx: u8,
+    hash_idx: u8,
+}
+
+#[allow(clippy::len_without_is_empty)]
+impl TocEntry {
+    /// Returns the type of the element this entry refers to.
+    ///
+    /// Returns `None` if the encoded element type is not known to Manticore.
+    pub fn element_type(&self) -> Option<ElementType> {
+        ElementType::from_wire_value(self.element_type)
+    }
+
+    /// Returns the format version for the element this entry refers to.
+    ///
+    /// Along with the type of the element, this value describes how to decode
+    /// the serialized element in the PFM.
+    pub fn format_version(&self) -> u8 {
+        self.format_version
+    }
+
+    /// Returns the flash address at which this entry's element begins.
+    ///
+    /// This value is measured from the start of the PFM's manifest frame,
+    /// rather than from the start of the PFM's table of contents.
+    pub fn offset(&self) -> u16 {
+        self.offset as _
+    }
+
+    /// Returns the length, in bytes, of this entry's element.
+    pub fn len(&self) -> u16 {
+        self.len as _
```

Comment:
These are both casts from u16 to u16? Did you want the return type to be usize for iteration's sake?

Reply:
Yeah I think that was supposed to be usize.